backend/src/main/java/hu/flowacademy/konyhatunder/service/RecipeService.java

```diff
@@ -97,6 +102,73 @@ public Recipe createRecipe(String stringRecipe, MultipartFile image) throws Json
+    }
+
+    public List<Recipe> listRecipesByIngredients(List<Ingredient> ingredientList) {
+        validateReceivedIngredients(ingredientList);
+        Set<Recipe> foundRecipes = new HashSet<>();
+        ingredientList.forEach(ingredient ->
+                foundRecipes.addAll(recipeRepository.findAllRecipesContainingIngredient(ingredient.getId())));
+        log.debug("Found {} recipe by criteria", foundRecipes.size());
+        return List.copyOf(foundRecipes);
+    }
+
+    public List<Recipe> listRecipesByCriteria(SearchByCriteriaDTO searchByCriteriaDTO) {
+        List<Recipe> foundRecipes = recipeRepository.findAll();
+        if (StringUtils.hasText(searchByCriteriaDTO.getName())) {
+            foundRecipes = foundRecipes.stream().filter(recipe -> recipe.getName().equalsIgnoreCase(searchByCriteriaDTO.getName())).collect(Collectors.toList());
+        }
+        if (searchByCriteriaDTO.getPreparationTime() != null) {
+            switch (searchByCriteriaDTO.getPreparationTime()) {
+                case 30:
+                    foundRecipes = foundRecipes.stream().filter(recipe ->
+                            recipe.getPreparationTime() < 30).collect(Collectors.toList());
+                    break;
+                case 59:
+                    foundRecipes = foundRecipes.stream().filter(recipe ->
+                            recipe.getPreparationTime() >= 30 && recipe.getPreparationTime() < 60).collect(Collectors.toList());
+                    break;
+                case 119:
+                    foundRecipes = foundRecipes.stream().filter(recipe ->
+                            recipe.getPreparationTime() >= 60 && recipe.getPreparationTime() < 120).collect(Collectors.toList());
+                    break;
+                case 179:
+                    foundRecipes = foundRecipes.stream().filter(recipe ->
+                            recipe.getPreparationTime() >= 120 && recipe.getPreparationTime() < 180).collect(Collectors.toList());
+                    break;
+                case 180:
+                    foundRecipes = foundRecipes.stream().filter(recipe ->
+                            recipe.getPreparationTime() >= 180).collect(Collectors.toList());
+                    break;
+                default:
+                    throw new ValidationException("Nem megfelelő elkészítési idő intervallum!");
+            }
+        }
+        if (StringUtils.hasText(searchByCriteriaDTO.getDifficulty())) {
+            foundRecipes = foundRecipes.stream().filter(recipe ->
+                    recipe.getDifficulty().equals(translateDifficulty(searchByCriteriaDTO.getDifficulty()))).collect(Collectors.toList());
+        }
+        if (searchByCriteriaDTO.getCategories() != null) {
+            foundRecipes = foundRecipes.stream().filter(recipe ->
+                    recipe.getCategories().stream().map(Category::getName).collect(Collectors.toList()).containsAll(searchByCriteriaDTO.getCategories())).collect(Collectors.toList());
```

Comment:
Hosszú a sor, és egy anyMatch()-el ki lehet iktatni a kozepso collect-et. De ezt is érdemes lenne levinni a query szintjére.

Reply:
Az a probléma, hogy kintről a cetegory list egy String lista, viszont a receptenek Category List-je van. Ezért kollekteltük ki előbb egy listába a categoryk neveit, hogy össze tudjuk hasonlítani őket.